src/frequenz/dispatch/actor.py

```diff
@@ -59,31 +45,97 @@ def __init__(
         self._running_state_change_sender = running_state_change_sender
-        self._poll_timer = Timer(poll_interval, SkipMissedAndDrift())
+        self._next_event_timer = Timer(
+            timedelta(seconds=100), SkipMissedAndResync(), auto_start=False
+        )
+        """The timer to schedule the next event.
+
+        Interval is chosen arbitrarily, as it will be reset on the first event.
+        """
+
+        self._scheduled_events: list[tuple[datetime, Dispatch]] = []
+        """The scheduled events, sorted by time.
+
+        Each event is a tuple of the scheduled time and the dispatch.
+        heapq is used to keep the list sorted by time, so the next event is
+        always at index 0.
+        """
 
     async def _run(self) -> None:
         """Run the actor."""
-        self._poll_timer.reset()
-        try:
-            async for _ in self._poll_timer:
-                await self._fetch()
-        except asyncio.CancelledError:
-            for task in self._scheduled.values():
-                task.cancel()
-            raise
+        # Initial fetch
+        await self._fetch()
+
+        stream = self._client.stream(microgrid_id=self._microgrid_id)
+
+        # Streaming updates
+        async for selected in select(self._next_event_timer, stream):
+            if selected_from(selected, self._next_event_timer):
+                if not self._scheduled_events:
+                    continue
+                _logger.debug(
+                    "Executing scheduled event: %s", self._scheduled_events[0][1]
+                )
+                await self._execute_scheduled_event(heappop(self._scheduled_events)[1])
+            elif selected_from(selected, stream):
+                _logger.debug("Received dispatch event: %s", selected.message)
+                dispatch = Dispatch(selected.message.dispatch)
+                match selected.message.event:
+                    case Event.CREATED:
+                        self._dispatches[dispatch.id] = dispatch
+                        await self._update_dispatch_schedule_and_notify(dispatch, None)
+                        await self._lifecycle_updates_sender.send(
+                            Created(dispatch=dispatch)
+                        )
+                    case Event.UPDATED:
+                        await self._update_dispatch_schedule_and_notify(
+                            dispatch, self._dispatches[dispatch.id]
+                        )
+                        self._dispatches[dispatch.id] = dispatch
+                        await self._lifecycle_updates_sender.send(
+                            Updated(dispatch=dispatch)
+                        )
+                    case Event.DELETED:
+                        self._dispatches.pop(dispatch.id)
+                        await self._update_dispatch_schedule_and_notify(None, dispatch)
+
+                        dispatch._set_deleted()  # pylint: disable=protected-access
```

Comment:
Also for a separate issue: if this function is used externally, then it should be public. If we want to restrict access only to certain users, one trick I've been using is to use either a `Protocol` for the public interface (where you don't expose the `set_deleted()` method) or just plain inheritance (`Dispatch` without `set_delete()` and a `DeletableDispatch` that inherits from it; internally you use `DeletableDispatch` but when returning it to the users, you return a plain `Dispatch`).

This actually made me think why are are storing state in a immutable object? That looks weird.
Another approach, since the `Dispatch`'s `deleted` `bool` and `running_state_change_synced` are only attached as mutable properties for internal **state** tracking, it might make more sense to separate the **state** from the immutable dispatch, and have a `DispatchState` **mutable** class and a separate `dict` to map dispatch ID to its state.

You can pass the state object to the `Dispatch` constructor and store it as a private property and expose *read-only* properties to access the state, if it needs to be accessed externally, and store the `dict` and mutate it from this actor only.

Reply:
* #57